File: Topics/Arrays&Hashing/two_sum/hash.cpp
// @Runtime 8 ms (93.65%)
// @Memory 10.7 MB (45.55%)

vector<int> twoSum(vector<int> &nums, int target)
{
    std::unordered_map<int, int> m;
    std::vector<int> v = {0, 0};
    int i = 0;

    for (const auto n : nums)
    {
        if (m.count(target - n))
        {
            v[0] = m[target - n];
            v[1] = i;
            return v;
        }
        else
            m[n] = i;
        ++i;
    }

    return v;
}

File: Topics/Arrays&Hashing/two_sum/hash.rs
// @Runtime 3 ms (50.98%)
// @Memory 2.5 MB (15.96%)

pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    use std::collections::HashMap;
    let mut map = HashMap::new();

    for (i, num) in nums.iter().enumerate() {
        let diff = target - num;

        if let Some(&j) = map.get(&diff) {
            return vec![j as i32, i as i32];
        }

        map.insert(num, i);
    }

    panic!("no solution");
}
